main.py

```diff
@@ -76,20 +78,26 @@ def __init__(self, connection):
         # self.plotMap(51.852667, -111.646972, self.radius, self.zoom)
         # self.plotMap(49.266430, -123.252162, self.radius, self.zoom)
 
-        # Init and connection of SerialThread and send and receive tasks
         self.printToConsole("Starting Connection")
-        self.SThread = SerialThread.SThread(connection)
-        self.SThread.sig_received.connect(self.receiveData)
-        self.sig_send.connect(self.SThread.queueMessage)
-        self.SThread.sig_print.connect(self.printToConsole)
-        self.SThread.start()
+
+        # Init and connection of ReadThread
+        self.ReadThread = ReadThread.ReadThread(connection)
+        self.ReadThread.sig_received.connect(self.receiveData)
+        self.ReadThread.sig_print.connect(self.printToConsole)
+        self.ReadThread.start()
+
+        # Init and connection of SendThread
+        self.SendThread = SendThread.SendThread(connection)
+        self.sig_send.connect(self.SendThread.queueMessage)
+        self.SendThread.sig_print.connect(self.printToConsole)
+        self.SendThread.start()
 
         # Mapping thread init and start
         thread = threading.Thread(target=self.threadLoop, daemon=True)
         thread.start()
 
     # Possible doc: receives from the serial thread loop of subpackets
-    def receiveData(self, dataBundle):  # TODO: ARE WE SURE THIS IS THREAD SAFE? USE QUEUE OR PUT IN SERIAL THREAD
+    def receiveData(self, dataBundle):
```

Comment:
For this part of data flow: What happens when the incoming data is dequeued through the new system and this is called? Will this (and the line below) properly queue up adding new data to RocketData regardless of the availability of the RocketData mutex lock?

Reply:
It will block the ReadThread on the next line until the mutex is free. But that's okay because its what that thread is for. We get a large advantage by implementing this thread and queue because we are essentially pipelining /parallelizing the CPU bound tasks with the IO bound tasks, so this allows each to work independently of each other allowing us to get ALL the data we can from the XBee. We could add multiple queues and processing threads at any point along the flow, but that would be parallelizing CPU bound tasks with other CPU bound tasks which is subject to diminishing returns and could actually make processing slower if we reach a CPU or IO throughput limit (especially if the threads are fighting for CPU time, but that's not the case for us, we are limited by IO/Xbee).

It would be interesting to monitor the size of the queues. If they are growing over time, then we know for sure that we need to improve the performance/parallelism of our data processing. But with how fast CPUs are these days and how slow the radio communication is, I'm pretty certain that the queue sizes would remain around 0-1 indicating that the IO/radio is the bottleneck.

It would be cool to maybe add a warning message if the queue is somehow getting too big.